### PR TITLE
Project submitted

### PR DESCRIPTION
**Pull request by nothingisfunny**

### CODE REVIEW

File: bin/tictactoe [hunk 1]
```diff
@@ -1,3 +1,25 @@
...
+		game = Game.new(player_1 = Player::Human.new("X"), player_2 = Player::Computer.new("O"), board = Board.new)
+		game.play
+	else input == "n"
+		game = Game.new(player_1 = Player::Computer.new("X"), player_2 = Player::Human.new("O"), board = Board.new)
+		game.play
+	end
+elsif input == "2"
+	game = Game.new(player_1 = Player::Human.new("X"), player_2 = Player::Human.new("O"), board = Board.new)
+	game.play
+end
```
Comment (randallreedjr) [inline bin/tictactoe]: Prefer to encapsulate CLI logic into a class. For example, the only code in your bin executable should be:

```ruby
#!/usr/bin/env ruby
       
require_relative '../config/environment'

TicTacToeCLI.new.call
```

And then you'd have a file: `lib/tic_tac_toe_cli.rb` with:

```ruby
class TicTacToeCLI
  def initialize
  end

  def call
   # Your game logic
  end
end
```

The code in your bin would generally just go in `call`.

File: lib/board.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
+class Board
+	attr_accessor :cells, :turns_taken
+	def initialize
+		@cells = Array.new(9, " ")
```
Comment (randallreedjr) [inline lib/board.rb]: I would just call `self.reset!` here and delegate to that method to create the default empty board.

File: lib/board.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+		return self.cells[input.to_i - 1]
+	end
+
+	def full?
+		self.cells.include?(" ") ? false : true
+	end
+
+	def turn_count
+		turns = 0
+		self.cells.each do |i|
```
Comment (randallreedjr) [inline lib/board.rb]: Use the highest level and most expressive iterator possible to solve a problem. [`count`](http://ruby-doc.org/core-2.2.0/Array.html#method-i-count) on the cells is perfect and the rest of this method is not needed. Entire body could read:

```ruby
def turn_count
  cells.count { |cell| cell == "X" || cell == "O" }
end
```

File: lib/board.rb [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+		end
+		return turns
+	end
+
+	def taken?(position)
+		self.cells[position.to_i-1] != " " 
+	end
+
+	def valid_move?(position)
+		!self.taken?(position) && position.to_i-1 >= 0 && position.to_i-1 <=9
```
Comment (randallreedjr) [inline lib/board.rb]: You can simplify checking the valid integer range, and make it more semantic, by using [`Comparable#between?`](http://ruby-doc.org/core-2.2.3/Comparable.html#method-i-between-3F).
`position.between?(1,9)`

File: lib/game.rb [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+		end
+	end
+
+	def over?
+		board.full? || self.won? || self.draw?
+	end
+
+	def won?
+			a = []
+			WIN_COMBINATIONS.each do |i|
```
Comment (randallreedjr) [inline lib/game.rb]: Here you can use [`Enumerable#detect`](http://ruby-doc.org/core-2.2.3/Enumerable.html#method-i-detect) to return the winning combination or `nil` if none is found.

```ruby
  def won?
    WIN_COMBINATIONS.find do |combo|
      @board.cells[combo[0]] == @board.cells[combo[1]] &&
      @board.cells[combo[0]] == @board.cells[combo[2]] &&
      @board.taken?(combo[0] + 1)
    end
  end
```

File: bin/tictactoe [hunk 6]
```diff
@@ -1,3 +1,25 @@
 #!/usr/bin/env ruby
 
 require_relative '../config/environment'
+puts "Welcome to Tic Tac Toe!"
+puts "press 0 for a 0 players game"
+puts "press 1 for a 1 players game"
+puts "press 2 for a 2 players game"
+input = gets.strip
+if input == "0"
+	game = Game.new(player_1 = Player::Computer.new("X"), player_2 = Player::Computer.new("O"), board = Board.new)
```
Comment (randallreedjr) [inline bin/tictactoe]: Being consistent with your whitespace, particularly using 2 spaces for indentation, will make your code more readable.

File: lib/players/human.rb [hunk 7]
```diff
@@ -0,0 +1,11 @@
+require_relative '../player.rb'
+
+class Player
+	class Human < Player
+		def move(board)
+			puts "Input 1-9:"
+			input = gets.strip
+			return input
```
Comment (randallreedjr) [inline lib/players/human.rb]: You don't need to store your return value in a local variable; just utilize the implicit return value of the expression.
```
def move(board)
  puts "Where would you like to move?"
  gets.chomp
end
```

File: lib/board.rb [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+		"-----------", 
+		" #{cells[6]} | #{cells[7]} | #{cells[8]} "
+	end
+
+	def position(input)
+		return self.cells[input.to_i - 1]
+	end
+
+	def full?
+		self.cells.include?(" ") ? false : true
```
Comment (randallreedjr) [inline lib/board.rb]: Avoid using the ternary operator to return `true` and `false`. Instead, rely on the implicit boolean return value of the expression. You can reverse (or negate) the boolean with the `!` (bang) operator.
```
def full?
  !self.cells.include?(" ")
end
```

File: lib/game.rb [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+		a = board.cells.select{|item| item == "X"}
+		b = board.cells.select{|item| item == "O"}
+		if self.won?
+			if a.length > b.length
+				return "X"
+		    else
+			    return "O"
+			end
+		else
+			nil
```
Comment (randallreedjr) [inline lib/game.rb]: You do not need to explicitly return `nil` from an `if` statement; an `if` will return `nil` if the condition fails.
```
if self.won?
  if a.length > b.length
    return "X"
  else
    return "O"
  end
end
```

**Comment by randallreedjr** [comment]
@nothingisfunny Great job! I left you a few notes, feel free to respond with any questions or comments. Sorry it took me so long to provide you with feedback, hopefully it’s still useful to you! Closing this PR for now.

**Comment by nothingisfunny** [comment]
@randallreedjr Hi, thanks for taking a look and leaving comments, they are really usefull! Honestly, I didn't expect that someone will check the code, and I still didn't finish improving the AI, so I'll work on code some more taking into account your notes.

**Pull request closed without merging**